fendermint/testing/materializer/src/arb.rs

```diff
@@ -0,0 +1,370 @@
+
+lazy_static! {
+    /// Assume we have as much tFIL on the root as the faucet would give.
+    static ref DEFAULT_BALANCE: Balance = Balance(TokenAmount::from_whole(100));
+}
+
+/// Select some items from a slice.
+fn choose_at_least<T: Clone>(g: &mut Gen, min_size: usize, xs: &[T]) -> Vec<T> {
+    let min_size = min(min_size, xs.len());
+
+    if min_size == xs.len() {
+        return Vec::from(xs);
+    }
+
+    // Say we have 10 items and we have 3 slots to fill.
+    //
+    // Imagine a simple algorithm that selects 1 item 3 times without replacement.
+    // Initially each item has 1/10 chance to be selected, then 1/9, then 1/8,
+    // but we would need to track which item has already been chosen.
+    //
+    // We want to do a single pass over the data.
+    //
+    // If we consider the 1st item, the chance that it doesn't get selected for any of the slots is:
+    // P_excl(1st) = 9/10 * 8/9 * 7/8 = 7/10
+    // P_incl(1st) = 1 - P_excl(1st) = 3/10
+    //
+    // So we roll the dice and with 30% probability we include the 1st item in the list.
+    //
+    // Then we have two cases to consider:
+    // 1. We included the 1st item, so now we have 2 slots and remaining 9 items to choose from.
+    //    P_incl(2nd | incl(1st)) = 1 - 8/9 * 7/8 = 1 - 7/9 = 2/9
+    // 2. We excluded the 1st item, so we still have 3 slots to fill and remaining 9 items to choose from.
+    //    P_incl(2nd | excl(1st)) = 1 - 8/9 * 7/8 * 6/7 = 1 - 6/9 = 3/9
+    //
+    // Thus, the probability of including each item is `remaining slots / remaining items`
+
+    let mut remaining_slots = min_size + usize::arbitrary(g) % (xs.len() - min_size);
+    let mut remaining_items = xs.len();
+
+    let mut chosen = Vec::new();
+    for x in xs {
+        if remaining_slots == 0 {
+            break;
+        }
+        if usize::arbitrary(g) % remaining_items < remaining_slots {
+            chosen.push(x.clone());
+            remaining_slots -= 1;
+        }
+        remaining_items -= 1;
+    }
+    chosen
+}
+
+fn choose_one<T: Clone>(g: &mut Gen, xs: &[T]) -> T {
+    g.choose(xs).expect("empty sequence to choose from").clone()
+}
+
+impl Arbitrary for ResourceId {
+    fn arbitrary(g: &mut Gen) -> Self {
+        let len = 3 + usize::arbitrary(g) % 6;
+
+        let id = (0..len)
+            .map(|_| {
+                let idx = usize::arbitrary(g) % RESOURCE_ID_CHARSET.len();
+                char::from(RESOURCE_ID_CHARSET[idx])
+            })
+            .collect();
+
+        Self(id)
+    }
+}
+
+impl Arbitrary for Balance {
+    fn arbitrary(g: &mut Gen) -> Self {
+        Self(Collateral::arbitrary(g).0)
+    }
+}
+
+impl Arbitrary for Manifest {
+    fn arbitrary(g: &mut Gen) -> Self {
+        gen_manifest(g, 4, 3, DEFAULT_BALANCE.clone())
+    }
+}
+
+fn gen_manifest(
+    g: &mut Gen,
+    max_children: usize,
+    max_level: usize,
+    default_balance: Balance,
+) -> Manifest {
+    let account_ids = (0..3 + usize::arbitrary(g) % 3)
```

Comment:
Should we supply the number of accounts externally?

Reply:
I don't think it makes too much a difference at this point. If we end up calling this with the intent that we want exactly N account, sure, but right now I just wanted some... I probably chose 3 because that is kind of the minimums of a non-standalone subnets with no tolerance for byzantine members, and then max 5 so we don't end up with no reuse, but like I say I don't use this for anything at the moment, it doesn't affect the general shape of the testnet like the other settings.